src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -191,109 +242,222 @@ public static int GetMaxDecodedFromUtf8Length(int length)
+                        goto InvalidExit;
+
+                    WriteThreeLowOrderBytes(bufferBytes + destIndex, i0);
+                    destIndex += 3;
+                }
+                else if (t2 != EncodingPad)
+                {
+                    int i2 = Unsafe.Add(ref decodingMap, (IntPtr)t2);
+
+                    i2 <<= 6;
+
+                    i0 |= i2;
+
+                    if (i0 < 0)
+                        goto InvalidExit;
+
+                    bufferBytes[destIndex] = (byte)(i0 >> 16);
+                    bufferBytes[destIndex + 1] = (byte)(i0 >> 8);
+                    destIndex += 2;
+                }
+                else
+                {
+                    if (i0 < 0)
+                        goto InvalidExit;
+
+                    bufferBytes[destIndex] = (byte)(i0 >> 16);
+                    destIndex += 1;
+                }
+
+            DoneExit:
+                bytesWritten = (int)destIndex;
+                return OperationStatus.Done;
+
+            InvalidExit:
+                bytesWritten = (int)destIndex;
+                return OperationStatus.InvalidData;
             }
-            else if (i2 != EncodingPad)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe void Avx2Decode(ref byte* srcBytes, ref byte* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, byte* destStart)
+        {
+            // The JIT won't hoist these "constants", so help it
+            Vector256<sbyte> lutHi = ReadVector<Vector256<sbyte>>(s_avxDecodeLutHi);
+            Vector256<sbyte> lutLo = ReadVector<Vector256<sbyte>>(s_avxDecodeLutLo);
+            Vector256<sbyte> lutShift = ReadVector<Vector256<sbyte>>(s_avxDecodeLutShift);
+            Vector256<sbyte> mask2F = s_avxDecodeMask2F;
+            Vector256<sbyte> shuffleConstant0 = Vector256.Create(0x01400140).AsSByte();
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = ReadVector<Vector256<sbyte>>(s_avxDecodeShuffleVec);
+            Vector256<int> permuteVec = ReadVector<Vector256<sbyte>>(s_avxDecodePermuteVec).AsInt32();
+
+            byte* src = srcBytes;
+            byte* dest = destBytes;
+
+            //while (remaining >= 45)
+            do
             {
-                i2 = Unsafe.Add(ref decodingMap, i2);
+                AssertRead<Vector256<sbyte>>(src, srcStart, sourceLength);
+                Vector256<sbyte> str = Avx.LoadVector256(src).AsSByte();
 
-                i2 <<= 6;
+                Vector256<sbyte> hiNibbles = Avx2.And(Avx2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
+                Vector256<sbyte> loNibbles = Avx2.And(str, mask2F);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHi, hiNibbles);
+                Vector256<sbyte> lo = Avx2.Shuffle(lutLo, loNibbles);
+                Vector256<sbyte> zero = Vector256<sbyte>.Zero;
```

Comment:
Does having a zero temp actually help at all?

Reply:
Maybe a little bit.
```asm
; ...
       C4411C57E4           vxorps   ymm12, ymm12, ymm12
       C4427500D2           vpshufb  ymm10, ymm1, ymm10
       C4412DDBD3           vpand    ymm10, ymm10, ymm11
       C4412D64D4           vpcmpgtb ymm10, ymm10, ymm12
; ...
```
As the `vxorps` comes some instructions before the next read of that registers, so it may break some RAW-dependency.

I'll try to hoist the `zero` before the loop. That would be better, if no spills occurs.

Reply:
Hm, I see some stack-spill from the ymm and xmm registers now in the dasm, that haven't been there before. Will investigate where they're from.

Reply:
The stack-spill got introduced in 31c4741ccb70007d393983ef38a37d99de715535
Affected is only decoding.

<details>
  <summary>dasm (excerpt) for decoding</summary>

```asm
G_M25157_IG05:
       498BC4               mov      rax, r12
       498BD5               mov      rdx, r13
       894DA4               mov      dword ptr [rbp-5CH], ecx
       448BD9               mov      r11d, ecx
       4C03D8               add      r11, rax
       4C895D80             mov      qword ptr [rbp-80H], r11
       418BF1               mov      esi, r9d
       4803F0               add      rsi, rax
       4183F918             cmp      r9d, 24
       0F8C4F030000         jl       G_M25157_IG11
       4889B578FFFFFF       mov      qword ptr [rbp-88H], rsi
       4C8D4ED3             lea      r9, [rsi-45]
       4C898D68FFFFFF       mov      qword ptr [rbp-98H], r9
       4D3BCC               cmp      r9, r12
       0F828C010000         jb       G_M25157_IG08
       448945A0             mov      dword ptr [rbp-60H], r8d
       48BF8B7D951D247F0000 mov      rdi, 0x7F241D957D8B
       C5FD1007             vmovupd  ymm0, ymmword ptr[rdi]
       C5FD118530FFFFFF     vmovupd  ymmword ptr[rbp-D0H], ymm0
       48BFCB7B951D247F0000 mov      rdi, 0x7F241D957BCB
       C5FD100F             vmovupd  ymm1, ymmword ptr[rdi]
       C5FD118D10FFFFFF     vmovupd  ymmword ptr[rbp-F0H], ymm1
       48BFEB7B951D247F0000 mov      rdi, 0x7F241D957BEB
       C5FD1017             vmovupd  ymm2, ymmword ptr[rdi]
       C5FD1195F0FEFFFF     vmovupd  ymmword ptr[rbp-110H], ymm2
       48BF08CBAAA2237F0000 mov      rdi, 0x7F23A2AACB08
       BE04000000           mov      esi, 4
       E8232E3E79           call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       48BFE80B008C237F0000 mov      rdi, 0x7F238C000BE8
       488B3F               mov      rdi, gword ptr [rdi]
       C5FD104708           vmovupd  ymm0, ymmword ptr[rdi+8]
       BF40014001           mov      edi, 0x1400140
       C5F96ECF             vmovd    xmm1, edi
       C4E27D58C9           vpbroadcastd ymm1, ymm1
       BF00100100           mov      edi, 0x11000
       C5F96ED7             vmovd    xmm2, edi
       C4E27D58D2           vpbroadcastd ymm2, ymm2
       48BF137E951D247F0000 mov      rdi, 0x7F241D957E13
       C5FD101F             vmovupd  ymm3, ymmword ptr[rdi]
       48BF5B7E951D247F0000 mov      rdi, 0x7F241D957E5B
       C5FD1027             vmovupd  ymm4, ymmword ptr[rdi]
       498BFC               mov      rdi, r12
       498BF5               mov      rsi, r13

G_M25157_IG06:
       C5FE6F2F             vmovdqu  ymm5, ymmword ptr[rdi]
       C5CD72D504           vpsrld   ymm6, ymm5, 4
       C5CDDBF0             vpand    ymm6, ymm6, ymm0
       C5D5DBF8             vpand    ymm7, ymm5, ymm0
       C57D108530FFFFFF     vmovupd  ymm8, ymmword ptr[rbp-D0H]
       C4623D00CE           vpshufb  ymm9, ymm8, ymm6
       C4412C57D2           vxorps   ymm10, ymm10, ymm10
       C57D109D10FFFFFF     vmovupd  ymm11, ymmword ptr[rbp-F0H]
       C4E22500FF           vpshufb  ymm7, ymm11, ymm7
       C4C145DBF9           vpand    ymm7, ymm7, ymm9
       C4C14564FA           vpcmpgtb ymm7, ymm7, ymm10
       C5FDD7C7             vpmovmskb eax, ymm7
       85C0                 test     eax, eax
       7577                 jne      SHORT G_M25157_IG07
       C5D574F8             vpcmpeqb ymm7, ymm5, ymm0
       C5C5FCF6             vpaddb   ymm6, ymm7, ymm6
       C5FD10BDF0FEFFFF     vmovupd  ymm7, ymmword ptr[rbp-110H]
       C4E24500F6           vpshufb  ymm6, ymm7, ymm6
       C5D5FCEE             vpaddb   ymm5, ymm5, ymm6
       C4E25504E9           vpmaddubsw ymm5, ymm5, ymm1
       C5D5F5EA             vpmaddwd ymm5, ymm5, ymm2
       C4E25500EB           vpshufb  ymm5, ymm5, ymm3
       C4E25D36ED           vpermd   ymm5, ymm4, ymm5
       C5FE7F2E             vmovdqu  ymmword ptr[rsi], ymm5
       4883C720             add      rdi, 32
       4883C618             add      rsi, 24
       488B8568FFFFFF       mov      rax, qword ptr [rbp-98H]
       483BF8               cmp      rdi, rax
       48898568FFFFFF       mov      qword ptr [rbp-98H], rax
       C57D118530FFFFFF     vmovupd  ymmword ptr[rbp-D0H], ymm8
       C57D119D10FFFFFF     vmovupd  ymmword ptr[rbp-F0H], ymm11
       C5FD11BDF0FEFFFF     vmovupd  ymmword ptr[rbp-110H], ymm7
       0F8657FFFFFF         jbe      G_M25157_IG06
```
</details>

Whe I add a static ctor to `Base64`, so the spill go away.
<details>
  <summary>dasm (excerpt) for decoding</summary>

```asm
G_M25157_IG05:
       4C8BF6               mov      r14, rsi
       4C8BF9               mov      r15, rcx
       458BE2               mov      r12d, r10d
       4D03E6               add      r12, r14
       8BD0                 mov      edx, eax
       4903D6               add      rdx, r14
       83F818               cmp      eax, 24
       0F8CFF010000         jl       G_M25157_IG11
       488D42D3             lea      rax, [rdx-45]
       483BC6               cmp      rax, rsi
       0F8201010000         jb       G_M25157_IG08
       49BE877D15AAD17F0000 mov      r14, 0x7FD1AA157D87
       C4C17D1006           vmovupd  ymm0, ymmword ptr[r14]
       49BFC77B15AAD17F0000 mov      r15, 0x7FD1AA157BC7
       C4C17D100F           vmovupd  ymm1, ymmword ptr[r15]
       49BEE77B15AAD17F0000 mov      r14, 0x7FD1AA157BE7
       C4C17D1016           vmovupd  ymm2, ymmword ptr[r14]
       49BEE80B0018D17F0000 mov      r14, 0x7FD118000BE8
       4D8B36               mov      r14, gword ptr [r14]
       C4C17D105E08         vmovupd  ymm3, ymmword ptr[r14+8]
       41BE40014001         mov      r14d, 0x1400140
       C4C1796EE6           vmovd    xmm4, r14d
       C4E27D58E4           vpbroadcastd ymm4, ymm4
       41BE00100100         mov      r14d, 0x11000
       C4C1796EEE           vmovd    xmm5, r14d
       C4E27D58ED           vpbroadcastd ymm5, ymm5
       49BE0F7E15AAD17F0000 mov      r14, 0x7FD1AA157E0F
       C4C17D1036           vmovupd  ymm6, ymmword ptr[r14]
       49BE577E15AAD17F0000 mov      r14, 0x7FD1AA157E57
       C4C17D103E           vmovupd  ymm7, ymmword ptr[r14]
       4C8BF6               mov      r14, rsi
       4C8BF9               mov      r15, rcx

G_M25157_IG06:
       C4417E6F06           vmovdqu  ymm8, ymmword ptr[r14]
       C4C13572D004         vpsrld   ymm9, ymm8, 4
       C535DBCB             vpand    ymm9, ymm9, ymm3
       C53DDBD3             vpand    ymm10, ymm8, ymm3
       C4427D00D9           vpshufb  ymm11, ymm0, ymm9
       C4411C57E4           vxorps   ymm12, ymm12, ymm12
       C4427500D2           vpshufb  ymm10, ymm1, ymm10
       C4412DDBD3           vpand    ymm10, ymm10, ymm11
       C4412D64D4           vpcmpgtb ymm10, ymm10, ymm12
       C4417DD7EA           vpmovmskb r13d, ymm10
       4585ED               test     r13d, r13d
       753C                 jne      SHORT G_M25157_IG07
       C53D74D3             vpcmpeqb ymm10, ymm8, ymm3
       C4412DFCC9           vpaddb   ymm9, ymm10, ymm9
       C4426D00C9           vpshufb  ymm9, ymm2, ymm9
       C4413DFCC1           vpaddb   ymm8, ymm8, ymm9
       C4623D04C4           vpmaddubsw ymm8, ymm8, ymm4
       C53DF5C5             vpmaddwd ymm8, ymm8, ymm5
       C4623D00C6           vpshufb  ymm8, ymm8, ymm6
       C4424536C0           vpermd   ymm8, ymm7, ymm8
       C4417E7F07           vmovdqu  ymmword ptr[r15], ymm8
       4983C620             add      r14, 32
       4983C718             add      r15, 24
       4C3BF0               cmp      r14, rax
       7692                 jbe      SHORT G_M25157_IG06
```
</details>

Found the issue: https://github.com/dotnet/corefx/blob/31c4741ccb70007d393983ef38a37d99de715535/src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs#L353
This is a reference to a static readonly field, so the JIT has to emit `CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE`, which unfortunately introduces the spills. So I need to avoid this static field access, by replacing with a ROS.